tests/rust/pipe-test/push_wait/client.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//======================================================================================================================
+// Imports
+//======================================================================================================================
+
+use ::anyhow::Result;
+use ::demikernel::{
+    demi_sgarray_t,
+    LibOS,
+    QDesc,
+    QToken,
+};
+use ::std::{
+    slice,
+    time::Duration,
+};
+use demikernel::runtime::types::demi_opcode_t;
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// Pipe client
+pub struct PipeClient {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Pipe name.
+    pipe_name: String,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+impl PipeClient {
+    /// Creates a new pipe client.
+    pub fn new(libos: LibOS, pipe_name: String) -> Result<Self> {
+        Ok(Self { libos, pipe_name })
+    }
+
+    // Runs the target pipe client.
+    pub fn run(&mut self) -> Result<()> {
+        let pipeqd: QDesc = self.libos.open_pipe(&format!("{}:rx", self.pipe_name))?;
+
+        // Allocate a scatter-gather array. Length and value are arbitrary.
+        let sga: demi_sgarray_t = match self.mksga(1, 123) {
+            Ok(sga) => sga,
+            Err(e) => {
+                // Clean up open pipe on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("mksga() failed (error={:?})", e)
+            },
+        };
+
+        // Push scatter-gather array.
+        let qt: QToken = match self.libos.push(pipeqd, &sga) {
+            Ok(qt) => qt,
+            Err(e) => {
+                // Clean up open pipe on error.
+                // Clean up scatter-gather array on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("push() failed (error={:?})", e)
+            },
+        };
+
+        // Succeed to release scatter-gather-array.
```

Comment:
This is not correct semantics, we should not release the scatter-gather army until after the wait on push returns either ok or error. I would recommend moving this match statement to after line 96.

Reply:
Well, this semantic is actually valid - it may not be correct from a design perspecive. Whenever we `push()`, we clone the scatter-gather array and bump the reference counter. We do it for all LibOSes, but here is the concerned code snippet for Catmem LibOS:

https://github.com/demikernel/demikernel/blob/f99757a9a8d10084cb8da579126b52ef46724e2d/src/rust/catmem/mod.rs#L176-L179

Reply:
Ah ok this is great, in our C++ code, this was not possible so we set the semantics in the C/C++ API accordingly

Reply:
Yeah.. We have shallow copies on the rust implementation :D